### PR TITLE
part of mixed training feature. enable TensorCore in many ops.

### PR DESCRIPTION
**Pull request by dzhwinter**

### CODE REVIEW

**Comment by dzhwinter** [comment]
last time, this PR cause float16 compile error, which has been fixed in https://github.com/PaddlePaddle/Paddle/pull/12756.

File: paddle/fluid/operators/math/cross_entropy.cu [hunk 1]
```diff
@@ -35,12 +49,12 @@ template <typename T>
...
                                        const int class_num) {
   int tid = threadIdx.x;
-  T val = 0;
+  T val(0);
 
   int idx = blockIdx.x * class_num + tid;
   int end = blockIdx.x * class_num + class_num;
   for (; idx < end; idx += blockDim.x) {
-    val += math::TolerableValue<T>()(std::log(X[idx])) * label[idx];
+    val += math::TolerableValue<T>()(log(X[idx])) * label[idx];
```
Comment (sneaxiy) [inline paddle/fluid/operators/math/cross_entropy.cu:62]: Maybe we can use `logf()` for `float32` and `log()` for `float64` here.

Comment (dzhwinter) [reply paddle/fluid/operators/math/cross_entropy.cu:62]: done.

File: paddle/fluid/operators/assign_value_op.cu.cc [hunk 2]
```diff
@@ -13,7 +13,10 @@ See the License for the specific language governing permissions and
...
 
 #include "paddle/fluid/operators/assign_value_op.h"
+#include "paddle/fluid/platform/float16.h"
 
 namespace ops = paddle::operators;
+namespace plat = paddle::platform;
 REGISTER_OP_CUDA_KERNEL(assign_value, ops::AssignValueKernel<int>,
-                        ops::AssignValueKernel<float>);
+                        ops::AssignValueKernel<float>,
+                        ops::AssignValueKernel<plat::float16>);
```
Comment (chengduoZH) [inline paddle/fluid/operators/assign_value_op.cu.cc:23]: `assign_value_op` should also support `double`.

File: paddle/fluid/operators/conv_cudnn_op.cu.cc [hunk 3]
```diff
@@ -127,27 +127,14 @@ class CUDNNConvOpKernel : public framework::OpKernel<T> {
...
-    // Tensor core is supported since the volta GPU and
-    // is only enabled when input and filter data are float16
-    if (dev_ctx.GetComputeCapability() >= 70 &&
-        std::type_index(typeid(T)) ==
-            std::type_index(typeid(platform::float16))) {
-      CUDNN_ENFORCE(platform::dynload::cudnnSetConvolutionMathType(
-          cudnn_conv_desc, CUDNN_TENSOR_OP_MATH));
-      // Currently tensor core is only enabled using this algo
+    if (platform::EnableTensorCore(dev_ctx, cudnn_conv_desc)) {
       algo = CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM;
```
Comment (chengduoZH) [inline paddle/fluid/operators/conv_cudnn_op.cu.cc:131]: As the cudnn doc saying, that Tensor Core Operations kernels will be triggered in these paths only when:
```
cudnnSetConvolutionMathType is called on the appropriate convolution descriptor setting mathType to CUDNN_TENSOR_OP_MATH.   
cudnnConvolutionForward is called using algo = CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM or CUDNN_CONVOLUTION_FWD_ALGO_WINOGRAD_NONFUSED; 
cudnnConvolutionBackwardData using algo = CUDNN_CONVOLUTION_BWD_DATA_ALGO_1 or CUDNN_CONVOLUTION_BWD_DATA_ALGO_WINOGRAD_NONFUSED; 
cudnnConvolutionBackwardFilter using algo = CUDNN_CONVOLUTION_BWD_FILTER_ALGO_1 or CUDNN_CONVOLUTION_BWD_FILTER_ALGO_WINOGRAD_NONFUSED.
```
And for algorithms other than *_ALGO_WINOGRAD_NONFUSED, the following are some of the requirements to run Tensor Core operations:
```
Input, Filter and Output descriptors (xDesc, yDesc, wDesc, dxDesc, dyDesc and dwDesc as applicable) have dataType = CUDNN_DATA_HALF.
The number of Input and Output feature maps is a multiple of 8.
The Filter is of type CUDNN_TENSOR_NCHW or CUDNN_TENSOR_NHWC. When using a filter of type CUDNN_TENSOR_NHWC, Input, Filter and Output data pointers (X, Y, W, dX, dY, and dW as applicable) need to be aligned to 128 bit boundaries.
```
Refer to https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#tensor_ops
So, although you set `algo` to `CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM ` here, the computation may not use the TensorCore.

File: paddle/fluid/operators/math/cross_entropy.h [hunk 4]
```diff
@@ -33,6 +35,21 @@ struct TolerableValue {
...
+template <>
+struct TolerableValue<float16> {
+  HOSTDEVICE float16 operator()(const float16& x) const {
+    if (isfinite(x))
+      return x;
+    else if (x > static_cast<float16>(0))
+      return std::numeric_limits<float16>::max();
+    else
+      return std::numeric_limits<float16>::min();
+  }
```
Comment (chengduoZH) [inline paddle/fluid/operators/math/cross_entropy.h:57]: Do you explain why `TolerableValue<float16>` is different with others here.

File: paddle/fluid/operators/softmax_cudnn_op.cu.cc [hunk 5]
```diff
@@ -78,4 +78,5 @@ REGISTER_OP_KERNEL(softmax, CUDNN, plat::CUDAPlace,
                    ops::SoftmaxCUDNNKernel<float>,
                    ops::SoftmaxCUDNNKernel<plat::float16>);
 REGISTER_OP_KERNEL(softmax_grad, CUDNN, plat::CUDAPlace,
-                   ops::SoftmaxGradCUDNNKernel<float>);
+                   ops::SoftmaxGradCUDNNKernel<float>,
+                   ops::SoftmaxGradCUDNNKernel<plat::float16>);
```
Comment (chengduoZH) [inline paddle/fluid/operators/softmax_cudnn_op.cu.cc:84]: `softmax` should also support double.

File: paddle/fluid/operators/math/cross_entropy.h (hunk 4 above)
Comment (dzhwinter) [reply paddle/fluid/operators/math/cross_entropy.h:57]: commented

File: paddle/fluid/operators/softmax_cudnn_op.cu.cc (hunk 5 above)
Comment (dzhwinter) [reply paddle/fluid/operators/softmax_cudnn_op.cu.cc:84]: done.

File: paddle/fluid/operators/assign_value_op.cu.cc (hunk 2 above)
Comment (dzhwinter) [reply paddle/fluid/operators/assign_value_op.cu.cc:23]: done.
